examples/files.rs

```diff
@@ -0,0 +1,218 @@
+        }
+
+        ratatui::widgets::Widget::render(Line::from(spans), area, buffer);
+    }
+}
+
+fn flatten_inner(
+    result: &mut Vec<Node<OsString>>,
+    open: &HashSet<Vec<OsString>>,
+    path: &Path,
+    current: &[OsString],
+) {
+    let Ok(read_dir) = path.read_dir() else {
+        return;
+    };
+
+    for entry in read_dir.flatten() {
+        let mut child_identifier = current.to_vec();
+        child_identifier.push(entry.file_name());
+
+        let is_dir = entry.metadata().is_ok_and(|metadata| metadata.is_dir());
+
+        result.push(Node {
+            identifier: child_identifier.clone(),
+            has_children: is_dir,
+            height: 1,
+        });
+
+        if open.contains(&child_identifier) {
+            flatten_inner(result, open, &entry.path(), &child_identifier);
+        }
+    }
+}
+
+struct App {
+    state: TreeState<OsString>,
+}
+
+impl App {
+    fn new() -> Self {
+        Self {
+            state: TreeState::default(),
+        }
+    }
+
+    fn draw(&mut self, frame: &mut Frame) {
+        let area = frame.size();
+        let data = FileTreeData(Path::new(".").to_owned());
+        let widget = Tree::new(&data)
+            .block(
+                Block::bordered()
+                    .title("File Tree")
+                    .title_bottom(format!("{:?}", self.state)),
+            )
+            .experimental_scrollbar(Some(
+                Scrollbar::new(ScrollbarOrientation::VerticalRight)
+                    .begin_symbol(None)
+                    .track_symbol(None)
+                    .end_symbol(None),
+            ))
+            .highlight_style(
+                Style::new()
+                    .fg(Color::Black)
+                    .bg(Color::LightGreen)
+                    .add_modifier(Modifier::BOLD),
+            )
+            .highlight_symbol(">> ");
+        frame.render_stateful_widget(widget, area, &mut self.state);
+    }
+}
+
+fn main() -> std::io::Result<()> {
+    // Terminal initialization
+    crossterm::terminal::enable_raw_mode()?;
+    let mut stdout = std::io::stdout();
+    crossterm::execute!(
+        stdout,
+        crossterm::terminal::EnterAlternateScreen,
+        crossterm::event::EnableMouseCapture
+    )?;
+    let mut terminal = Terminal::new(CrosstermBackend::new(stdout))?;
+
+    // App
+    let app = App::new();
+    let res = run_app(&mut terminal, app);
+
+    // restore terminal
+    crossterm::terminal::disable_raw_mode()?;
+    crossterm::execute!(
+        terminal.backend_mut(),
+        crossterm::terminal::LeaveAlternateScreen,
+        crossterm::event::DisableMouseCapture
+    )?;
+    terminal.show_cursor()?;
+
+    if let Err(err) = res {
+        println!("{err:?}");
+    }
+
+    Ok(())
+}
+
+fn run_app<B: Backend>(terminal: &mut Terminal<B>, mut app: App) -> std::io::Result<()> {
+    const DEBOUNCE: Duration = Duration::from_millis(20); // 50 FPS
```

Comment:
Is this necessary for the example? It might over complicate things a bit more than needed.

Reply:
It’s the same for all examples and especially without these traits it’s really effective for performance especially on scrolling.

could be moved into a common example base. It seems like too much for this repo so maybe a crate or something behind a feature flag on ratatui? Definitely no perfect solution here.